### PR TITLE
Rename RangeArgument to RangeBounds, move it and Bound to libcore

### PR DESCRIPTION
**Pull request by SimonSapin**
As proposed in the tracking issue: https://github.com/rust-lang/rust/issues/30877

Changes to *stable* items:

* `core::ops::Bound` / `std::ops::Bound` is new
* `std::collections::Bound` is a deprecated reexport of it (does this actually cause a warning?)

Changes to *unstable* items

* `alloc::Bound` is gone
* `alloc::range::RangeArgument` is moved to `core::ops::RangeBounds` / `std::ops::RangeBounds`
* `alloc::range` is gone
* `std::collections::range::RangeArgument` is deprecated reexport, to be removed later
* `std::collections::range` is deprecated, to be removed later
* `impl RangeBounds<T> for Range{,From,To,Inclusive,ToInclusive}<&T>` are added

The idea of replacing this trait with a type to be used with `Into<_>` is left for future consideration / work.

(Fixes https://github.com/rust-lang-nursery/rust-clippy/issues/2552.)

### CODE REVIEW

**Comment by SimonSapin** [comment]
I tried adding this:

```rust
impl<'a, R, T: ?Sized + 'a> RangeBounds<T> for R where R: RangeBounds<&'a T> {
    // …
}
```

and got:

```rust
error[E0275]: overflow evaluating the requirement `ops::range::RangeFull: ops::range::RangeBounds<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&_>`
  |
  = help: consider adding a `#![recursion_limit="128"]` attribute to your crate
  = note: required because of the requirements on the impl of `ops::range::RangeBounds<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&_>` for `ops::range::RangeFull`
  = note: required because of the requirements on the impl of `ops::range::RangeBounds<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&_>` for `ops::range::RangeFull`
[…]
  = note: required because of the requirements on the impl of `ops::range::RangeBounds<&&_>` for `ops::range::RangeFull`
  = note: required because of the requirements on the impl of `ops::range::RangeBounds<&_>` for `ops::range::RangeFull`

error: aborting due to previous error
```

**Comment by clarcharr** [comment]
Oh, huh. I'm a bit confused why that's failing, considering how `AsRef` has a similar implementation…

**Comment by SimonSapin** [comment]
`AsRef` has an impl `for &T`, not `for T where /*…*/`. The latter is more general and has more potential overlap with other impls.

**Comment by SimonSapin** [comment]
Added `impl RangeBounds<T> for Range{,From,To,Inclusive,ToInclusive}<&T>`

**Comment by SimonSapin** [comment]
@clarcharr You comment https://github.com/rust-lang/rust/pull/44518#discussion_r139056204 suggests that the thing with ranges of strings is not just adding `&` to some types, but also `String` v.s. `str`.

I’ve just pushed another commit that generalizes the impls to:

```rust
impl<T: ?Sized> RangeBounds<T> for RangeFull
impl<T: ?Sized, U: Borrow<T>> RangeBounds<T> for RangeFrom<U>
impl<T: ?Sized, U: Borrow<T>> RangeBounds<T> for RangeTo<U>
impl<T: ?Sized, U: Borrow<T>> RangeBounds<T> for Range<U>
impl<T: ?Sized, U: Borrow<T>> RangeBounds<T> for RangeInclusive<U>
impl<T: ?Sized, U: Borrow<T>> RangeBounds<T> for RangeToInclusive<U>
impl<T: ?Sized, U: Borrow<T>> RangeBounds<T> for (Bound<U>, Bound<U>)
```

Does this seem like a good idea?

**Comment by SimonSapin** [comment]
(As a reminder: `T` and `&T` both implement `Borrow<T>`, only the latter implements `AsRef<T>`. Trying to have two sets of impls for `Range*<T>` and `Range*<U> where U: AsRef<T>` causes an overlap error because some type could implement `AsRef<Foo> for Foo`.)

**Comment by SimonSapin** [comment]
Ah, this was probably the issue you meant:

```rust
---- btree/map.rs - btree::map::BTreeMap<K, V>::range_mut (line 818) stdout ----
	error[E0283]: type annotations required: cannot resolve `_: std::cmp::Ord`
 --> btree/map.rs:824:25
  |
9 | for (_, balance) in map.range_mut("B".."Cheryl") {
  |                         ^^^^^^^^^

thread 'rustc' panicked at 'couldn't compile the test', librustdoc/test.rs:296:13
note: Run with `RUST_BACKTRACE=1` for a backtrace.


failures:
    btree/map.rs - btree::map::BTreeMap<K, V>::range_mut (line 818)
```

I think that code has two possible concrete types for some type parameter. Both would end up producing the same behavior, but the compiler is still unhappy about the ambiguity.

Undoing that commit.

**Comment by clarcharr** [comment]
Ah, yeah, that's why I could never complete the original PR.

It's truly unfortunate that `RangeArgument` exists the way it does, as it can't really be changed despite being unstable. It'd be better if we could return the start and end by value, as `Into<RangeBounds>` would, but that would result in different behaviour and not-so-subtle differences from what we've got right now.

I guess that the best option is to just stabilise it as-is and deal with what we've got. I appreciate the name change and the effort, though!

File: src/libstd/collections/mod.rs [hunk 1]
```diff
@@ -420,7 +420,8 @@
 #![stable(feature = "rust1", since = "1.0.0")]
 
 #[stable(feature = "rust1", since = "1.0.0")]
-pub use alloc::Bound;
+#[rustc_deprecated(reason = "moved to `std::ops::Bound`", since = "1.26.0")]
```
Comment (alexcrichton) [inline src/libstd/collections/mod.rs]: We may want to also tag this with `#[doc(hidden)]` to prevent it showing up in the docs

**Review by alexcrichton — COMMENTED** [review]
Looks great to me, thanks @SimonSapin!

**Comment by alexcrichton** [comment]
@rfcbot fcp merge

Since this is touching and moving stable items I'd like to get some more eyes on this, but I'm not expecting anything surprising here

**Comment by SimonSapin** [comment]
If we’re doing the FCP dance, should we also propose to stabilize? If we decide on a dedicated trait (over `Into<_>`) this PR fixes the remaining issues that I could find in the tracking issue.

**Comment by alexcrichton** [comment]
I'd personally prefer to land this first and then have a separate discussion about the various bounds/impls/such

File: src/libstd/collections/mod.rs (hunk 1 above)
Comment (SimonSapin) [reply src/libstd/collections/mod.rs]: Sounds good, done.

**Comment by kennytm** [comment]
This has broken clippy's `ui/needless_pass_by_value.rs` test. It has printed nothing.

**Comment by SimonSapin** [comment]
I can reproduce this but I have no idea why it happens :(

**Comment by Manishearth** [comment]
Rebased and fixed. I'll deal with getting the clippy stuff landed correctly.

@bors r=alexcrichton

**Comment by Manishearth** [comment]
Unsure why that didn't work, but it let it land, so that's good enough for now. Will fix later.

**Pull request merged**